src/hotspot/share/oops/resolvedFieldEntry.hpp

```diff
@@ -0,0 +1,144 @@
+// a "field index", which is an index into the array of ResolvedFieldEntry.
+
+//class InstanceKlass;
+class ResolvedFieldEntry {
+  friend class VMStructs;
+
+  InstanceKlass* _field_holder; // Field holder klass
+  int _field_offset;            // Field offset in bytes
+  u2 _field_index;              // Index into field information in holder InstanceKlass
+  u2 _cpool_index;              // Constant pool index
+  u1 _tos;                      // TOS state
+  u1 _flags;                    // Flags: [0000|00|is_final|is_volatile]
+  u1 _get_code, _put_code;      // Get and Put bytecodes of the field
+
+public:
+  ResolvedFieldEntry(u2 cpi) :
+    _field_holder(nullptr),
+    _field_offset(0),
+    _field_index(0),
+    _cpool_index(cpi),
+    _tos(0),
+    _flags(0),
+    _get_code(0),
+    _put_code(0) {}
+  ResolvedFieldEntry() :
+    ResolvedFieldEntry(0) {}
+
+  // Bit shift to get flags
+  // Note: Only two flags exists at the moment but more could be added
+  enum {
+      is_volatile_shift     = 0,
+      is_final_shift        = 1, // unused
+  };
+
+  // Getters
+  InstanceKlass* field_holder() const { return _field_holder; }
+  int field_offset()            const { return _field_offset; }
+  u2 field_index()              const { return _field_index;  }
+  u2 constant_pool_index()      const { return _cpool_index;  }
+  u1 tos()                      const { return _tos;          }
+  u1 get_code()                 const { return Atomic::load_acquire(&_get_code);      }
+  u1 put_code()                 const { return Atomic::load_acquire(&_put_code);      }
+  bool is_final()               const { return (_flags & (1 << is_final_shift))    != 0; }
+  bool is_volatile ()           const { return (_flags & (1 << is_volatile_shift)) != 0; }
+  bool is_resolved(Bytecodes::Code code) const {
+    switch(code) {
+    case Bytecodes::_getstatic:
+    case Bytecodes::_getfield:
+      return (get_code() == code);
+    case Bytecodes::_putstatic:
+    case Bytecodes::_putfield:
+      return (put_code() == code);
+    default:
+      ShouldNotReachHere();
+      return false;
+    }
+  }
+
+  // Printing
+  void print_on(outputStream* st) const;
+
+  void set_flags(bool is_final, bool is_volatile) {
+    u1 new_flags = (static_cast<u1>(is_final) << static_cast<u1>(is_final_shift)) | static_cast<u1>(is_volatile);
+    _flags = new_flags;
+  }
+
+  inline void set_bytecode(u1* code, u1 new_code) {
+  #ifdef ASSERT
+    // Read once.
+    volatile Bytecodes::Code c = (Bytecodes::Code)*code;
+    assert(c == 0 || c == new_code || new_code == 0, "update must be consistent");
+  #endif
+    Atomic::release_store(code, new_code);
+  }
+
+  // Populate the strucutre with resolution information
+  void fill_in(InstanceKlass* klass, intx offset, int index, int tos, u1 b1, u1 b2) {
+    _field_holder = klass;
+    _field_offset = offset;
+    _field_index = index;
+    _tos = tos;
+
+    // This has to be done last
```

Comment:
Maybe better use `OrderAccess::release()` here instead of 2x `Atomic::release_store` above?

Reply:
Is the comment accurate? The semantics would be different. Does seeing the update to `_put_code` require that the update to `_get_code` is also seen?

Reply:
The comment seems to be unclear so I will update it. The fields `_get_code` and `_put_code` do not require the other to be seen, they just have to be set after all of the other fields.